pkg/cmd/roachtest/tests/lease.go

```diff
@@ -0,0 +1,187 @@
+// unavailability following a leaseholder crash. It tests the simplest possible
+// failure:
+//
+//   - A process crash, where the host/OS remains available (in particular, the
+//     TCP/IP stack is responsive and sends immediate RST packets to peers).
+//
+//   - No system ranges located on the crashed node.
+//
+//   - SQL clients do not connect to the crashed node.
+//
+//   - The workload consists of individual point reads and writes.
+//
+// We expect the worst-case (pMax) latency to be slightly larger than the lease
+// interval (9s), to account for lease acquisition and retry latencies. We do
+// not assert this, but instead export latency histograms for graphing.
+//
+// The cluster layout is as follows:
+//
+// n1-n3: System ranges and SQL gateways.
+// n4-n6: Workload ranges.
+// n7:    Workload runner.
+//
+// The test runs a kv50 workload with batch size 1, using 256 concurrent
+// workers directed at n1-n3. n4-n6 are killed and restarted in order, with 30
+// seconds between each operation, for 3 cycles totaling 9 crashes.
+func runAcquireLeaseCrashNonSystem(ctx context.Context, t test.Test, c cluster.Cluster) {
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	c.Put(ctx, t.DeprecatedWorkload(), "./workload")
+
+	// Create cluster with one rack locality per node and configure it.
+	require.Equal(t, 7, c.Spec().NodeCount)
+
+	opts := option.DefaultStartOpts()
+	settings := install.MakeClusterSettings(install.NumRacksOption(c.Spec().NodeCount))
+	c.Start(ctx, t.L(), opts, settings, c.Range(1, 6))
+
+	conn := c.Conn(ctx, t.L(), 1)
+
+	t.Status("configuring cluster")
+	_, err := conn.ExecContext(ctx, `SET CLUSTER SETTING kv.range_split.by_load_enabled = 'false'`)
+	require.NoError(t, err)
+
+	// Constrain all existing zone configs to n1-n3.
+	conn2 := c.Conn(ctx, t.L(), 1)
+	rows, err := conn.QueryContext(ctx, `SELECT target FROM [SHOW ALL ZONE CONFIGURATIONS]`)
+	require.NoError(t, err)
+	for rows.Next() {
+		var target string
+		require.NoError(t, rows.Scan(&target))
+		_, err = conn2.ExecContext(ctx, fmt.Sprintf(
+			`ALTER %s CONFIGURE ZONE USING num_replicas = 3, constraints = '[-rack=3, -rack=4, -rack=5]'`,
+			target))
+		require.NoError(t, err)
+	}
+	require.NoError(t, rows.Err())
+	rows.Close()
+	require.NoError(t, conn2.Close())
+
+	// Wait for upreplication.
+	require.NoError(t, WaitFor3XReplication(ctx, t, conn))
+
+	// Create the kv database, constrained to n4-n6. Despite the zone config, the
+	// ranges will initially be distributed across all cluster nodes.
+	t.Status("creating workload database")
+	_, err = conn.ExecContext(ctx, `CREATE DATABASE kv`)
+	require.NoError(t, err)
+	_, err = conn.ExecContext(ctx, `ALTER DATABASE kv CONFIGURE ZONE USING `+
+		`num_replicas = 3, constraints = '[-rack=0, -rack=1, -rack=2]'`)
+	require.NoError(t, err)
+	c.Run(ctx, c.Node(7), `./workload init kv --splits 1000 {pgurl:1}`)
+
+	// The replicate queue takes forever to move the kv ranges from n1-n3 to
+	// n4-n6, so we do it ourselves.
```

Comment:
Wouldn't it be faster to do this first, before creating 1000 ranges?

Reply:
You would think so, but `workload init kv --splits` command sprays the ranges across all nodes regardless of the initial range layout. Tried a bunch of stuff to make it position them correctly without any range movement, and didn't want to start messing around with patching `workload`.